sycl/test-e2e/KernelAndProgram/free_function_kernels.cpp

```diff
@@ -0,0 +1,274 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+// UNSUPPORTED: hip
+
+// This test tests free function kernel code generation and execution.
+
+#include <iostream>
+#include <sycl/sycl.hpp>
+
+using namespace sycl;
+
+// Kernel finder.
+class KernelFinder {
+  queue &Queue;
+  std::vector<kernel_id> AllKernelIDs;
+
+public:
+  KernelFinder(queue &Q) : Queue(Q) {
+    // Obtain kernel bundle
+    kernel_bundle Bundle =
+        get_kernel_bundle<bundle_state::executable>(Queue.get_context());
+    std::cout << "Bundle obtained\n";
+    AllKernelIDs = get_kernel_ids();
+    std::cout << "Number of kernels = " << AllKernelIDs.size() << std::endl;
+    for (auto K : AllKernelIDs) {
+      std::cout << "Kernel obtained: " << K.get_name() << std::endl;
+    }
+    std::cout << std::endl;
+  }
+
+  kernel get_kernel(const char *name) {
+    kernel_bundle Bundle =
+        get_kernel_bundle<bundle_state::executable>(Queue.get_context());
+    for (auto K : AllKernelIDs) {
+      auto Kname = K.get_name();
+      if (strcmp(name, Kname) == 0) {
+        kernel Kernel = Bundle.get_kernel(K);
+        return Kernel;
+      }
+    }
+    std::cout << "No kernel named " << name << " found\n";
+    exit(1);
+  }
+};
+
+void printUSM(int *usmPtr, int size) {
+  std::cout << "usmPtr[] = {";
+  for (int i = 0; i < size; i++) {
+    std::cout << usmPtr[i] << ", ";
+  }
+  std::cout << "}\n";
+}
+
+bool checkUSM(int *usmPtr, int size, int *Result) {
+  bool Pass = true;
+  for (int i = 0; i < size; i++) {
+    if (usmPtr[i] != Result[i]) {
+      Pass = false;
+      break;
+    }
+  }
+  if (Pass)
+    return true;
+
+  std::cout << "Expected = {";
+  for (int i = 0; i < size; i++) {
+    std::cout << Result[i] << ", ";
+  }
+  std::cout << "}\n";
+  std::cout << "Result   = {";
+  for (int i = 0; i < size; i++) {
+    std::cout << usmPtr[i] << ", ";
+  }
+  std::cout << "}\n";
+  return false;
+}
+
+SYCL_EXTERNAL
+SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
+    (ext::oneapi::experimental::single_task_kernel))
+void ff_0(int *ptr, int start, int end) {
+  for (int i = start; i <= end; i++)
+    ptr[i] = start + end;
+}
+
+bool test_0(queue Queue, KernelFinder &KF) {
+  constexpr int Range = 10;
+  int *usmPtr = malloc_shared<int>(Range, Queue);
+  int start = 3;
+  int end = 5;
+  int Result[Range] = {0, 0, 0, 8, 8, 8, 0, 0, 0, 0};
+  range<1> R1{Range};
+
+  memset(usmPtr, 0, Range * sizeof(int));
+  Queue.submit([&](handler &Handler) {
+    Handler.single_task([=]() {
+      for (int i = start; i <= end; i++)
+        usmPtr[i] = start + end;
+    });
+  });
+  Queue.wait();
+  bool PassA = checkUSM(usmPtr, Range, Result);
+  std::cout << "Test 0a: " << (PassA ? "PASS" : "FAIL") << std::endl;
+
+  kernel Kernel = KF.get_kernel("_Z18__sycl_kernel_ff_0Piii");
```

Comment:
This test isn't excluded on Windows, but I would expect the kernel lookup to fail there since an IA64 mangled name is supplied. Or are IA64 mangled names being used for free function kernels on Windows too? (If I recall correctly, IA64 mangled names are currently generated for Windows for regular SYCL functions, right?)

Reply:
Yes, kernel names are not host-OS dependent and this test will pass on Windows.